### PR TITLE
hwdb: Add new snd_hda_intel power-saving blacklist hwdb file

### PR DESCRIPTION
**Pull request by jwrdegoede**
OpenSUSE has already been shipping kernels with
CONFIG_SND_HDA_POWER_SAVE_DEFAULT=1 for a while now and Fedora is going
to do the same starting with Fedora 28. In general this is a good thing
as it saves aprox. 0.4W of power when the audio codec is not active.

But on some devices with snd_hda_intel audio, this leads to a plopping or
crackling sound each time the codec is turned on/off.

This usually points to a problem in the hda-codec driver for the specific
external codec used together with the Intel HDA audio integrated into the
chipset. Unfortunately there are many different external codecs and even
more ways these could be wired up and then they can be paired with an
external amplifier which may have multiple enable pins, etc. So fixing
this is a tedious and time-consuming process which needs to be done per
board.

TL;DR: the undesired sounds when powersaving kicks in are typically a
kernel-bug but getting these fixed in a timely manner on all affected
models is not feasible.

Since CONFIG_SND_HDA_POWER_SAVE_DEFAULT=1 works problem free on aprox
95-99% of all models out there and since it gives a nice increase in
battery life we really want to have it on by default

So this commit implements plan B for devices on which this causes problems,
a blacklist in hwdb which turns of the power-saving on these specific
devices.

This initial commit adds only one model laptop to the hwdb, but I expect
more models to follow in the future.

Related bugzillas:
https://bugzilla.kernel.org/show_bug.cgi?id=198611
https://bugzilla.redhat.com/show_bug.cgi?id=1525561
https://bugzilla.redhat.com/show_bug.cgi?id=1525104
https://bugzilla.redhat.com/show_bug.cgi?id=1526678
https://bugzilla.redhat.com/show_bug.cgi?id=1533858

### CODE REVIEW

**Comment by poettering** [comment]
Hmm. So far the quirks database was mostly about hw quirks, not driver quirks. With this you are entering entirely new territory here, and I am not sure I like this very much, because software should really be fixed rather than worked around, and if we carry a quirk we should at least have a clear philosophy in place how these entries are removed later on. I mean, we really don't want to collect information about lots of obsolete driver versions here, do we?

Moreover, how do you intend to handle dealing with kernels that are updated and fix this cases? suddenly our hwdb would become kernel-version dependent, we'd have to have a strategy for that, no?

File: rules/90-hda-powersave.rules [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="hda_powersave_end"
+
+SUBSYSTEM=="sound", DRIVERS=="snd_hda_intel", \
+  IMPORT{builtin}="hwdb 'snd_hda_intel:$attr{subsystem_vendor}:$attr{subsystem_device}:$attr{[dmi/id]modalias}'"
+
+ENV{SND_HDA_INTEL_POWERSAVE}!="", \
+  RUN+="/bin/sh -c 'echo $env{SND_HDA_INTEL_POWERSAVE} > /sys/module/snd_hda_intel/parameters/power_save'"
```
Comment (poettering) [inline rules/90-hda-powersave.rules:9]: hrmpf.. so we alter per-driver switches based on devices found. This means a system with two hda sound cards will get quite an undeterministic behaviour here, no?

**Comment by jwrdegoede** [comment]
Hi Lennart,

You're right I did not properly think through how to remove entries once the kernel gets fixed. My main reason for doing this through hwdb is that typically the upstream kernel maintainers are against high-maintenance quirk tables in the kernel and prefer hwdb for this. But I see that the sound subsys and esp. the hda driver is full of quirks, so adding another quirk table for this there will probably be accepted and will allow easy removing of the disable-power-save-quirk at the same time as a codec driver gets fixed (or more likely a codec driver gets a quirk fixing this with a smaller-hammer approach).

Anyways this is probably best dealt with in the kernel, so lets close this.

Regards,

Hans
